File: 001601-001800/001710.卡车上的最大单元数/1710.go
package main

import (
	"sort"
)

// 方法一：用时 20 ms
// 执行用时超过了 84 %
// 消耗内存超过了 95 %
func maximumUnits(boxTypes [][]int, truckSize int) int {
	sort.Slice(boxTypes, func(i, j int) bool {
		return boxTypes[j][1] < boxTypes[i][1]
	})
	ret := 0
	for i := 0; i < len(boxTypes) && truckSize > 0; i++ {
		curSize := boxTypes[i][0]
		if boxTypes[i][0] > truckSize {
			curSize = truckSize
		}
		ret += curSize * boxTypes[i][1]
		truckSize -= curSize
	}
	return ret
}

func main ()  {
	//fmt.Println(maximumUnits([]int{}, 8))
	//fmt.Println(maximumUnits([]int{}, 8))
}

File: 001601-001800/001710.卡车上的最大单元数/1710.js
/**
 * @param {number[][]} boxTypes
 * @param {number} truckSize
 * @return {number}
 */
 // 方法一：用时 76 ms
 // 执行用时超过了 92 %
 // 消耗内存超过了 42 %
var maximumUnits = function(boxTypes, truckSize) {
  boxTypes = boxTypes.sort((a, b) => (b[1] - a[1]))
  let ret = 0
  for (let i = 0; i < boxTypes.length && truckSize > 0; i++) {
    let curSize = boxTypes[i][0]
    if (boxTypes[i][0] > truckSize) {
      curSize = truckSize
    }
    ret += curSize * boxTypes[i][1]
    truckSize -= curSize
  }
  return ret
};

console.log(maximumUnits([[1, 3], [2, 2], [3, 1]], 4))
console.log(maximumUnits([[5, 10], [2, 5], [4, 7], [3, 9]], 10))
